docs/developers/talawa-api/error-handling.md

```diff
@@ -0,0 +1,625 @@
+                        minimumLength: 8
+                   })
+           }
+           // Approach when we need to fall on a general error based on the interface contract `UserError`
+           If (CERTAIN CHECK WHERE WE WOULD NEED TO ADD THAT ERROR IN THE 
+                  signUpErrors ARRAY) {
+                    userErrors.push({
+                            __typename:"UserError" ,
+                            message: "message" , 
+                            path: "path"
+                    })
+                  
+            }
+
+          // Here we will be returning multiple errors in the form on an array of signUpErrors
+          if (signUpErrors) {
+               return  {
+                   signUpData: {
+                         user : userObj,
+                          .... other AuthData fields
+                   } , 
+                   signUpErrors
+
+               }
+          }
+          
+          EVERYTHING IS OKAY, CREATE THE USER IN DB,
+          
+          CreatedUserObj = DB.CREATE(USER);
+
+          return  {
+                   signUpData: {
+                         user : CreatedUserObj,
+                          .... other AuthData fields
+                   } , 
+                   signUpErrors
+
+           }
+
+     }
+  }
+}
+```
+
+In this approach for resolving Field Errors ->
+
+1. Multiple Errors can be sent back to the client when needed in case the client apps need to display those errors concurrently with UI elements
+2. An Error for a field means only that field will be sent as null in the query, other fields will remain unaffected.
+3. A general purpose interface contract `UserError` to fall back to in case we want to send a general purpose error back to the client.
+
+
+#### Client
+
+Let us take a look at how the clients would be making this query.
+
+```javascript
+mutation {
+  signUp(
+    input: {
+      firstName: "Harry"
+      lastName: "Potter"
+      email: "someDuplicateEmail"
+      password: "12345"
+    }
+    ) {
+    signUpData { 
+      user: {
+        _id,
+        firstName,
+        lastName,
+        email
+      } ,
+      accessToken,
+      refreshToken
+     }
+    signUpErrors {
+      # Specific cases
+      ... on EmailTaken {
+        __typename
+        message
+        path
+        suggestion
+      }
+
+      ... on PasswordTooShort {
+        __typename
+        message
+        path
+        minimumLength
+      }
+
+      # Interface contract to handle general purpose Error
+      ... on UserError {
+        message
+        path
+      }
+    }
+  }
+}
+```
+
+As you can see from this mutation input, the following mutation should result in the `EmailTaken` and `PasswordTooShort` Error based on our resolver.
+
+The UserError is to handle any general purpose error that way we get extensibility of the interface with the expressivity of the union.
+
+Hence, this mutation will return a response that looks like this.
+
+```rust
+
+data: {
```

Comment:
This response is incorrect. If email is taken or password is not according to constraints the returned user object should be null. The user object you're returning is to return the user you created, here no user will be created because the constraints are never fullfilled. This is not an example of partial errors.

Reply:
@xoldyckk Yes you are right. Although that example was meant as an example of field errors primarily. Will correct that response and remove mention of partial data there. 

So basically the nested resolver nullability and partial data concept are basically the same, right? So should I mention that in the nested resolver error section ? or make a new example with another connection for that (although I doubt it would be necessary)?

Reply:
Also regarding the scalar fields I think we do that within the parent resolver itself since then it will be easier to query for the clients since all errors would be in one place? What do you think?

Reply:
@kb-0311 Think of nested resolver as just a field. The field can be scalar or a complex objects. Whether you choose to implement a seperate resolver for scalar fields or not depends on your implementation.

Since, you return most of the scalars in the parent resolver itself graphql doesn't throw out errors to resolve the fields which are asked for in schema. This is because the fields in schema and fields in database have the same name and data type.

Same is not true for custom scalar and complex object fields. Example for custom scalar is `postsCount`, there's no database field named `postCount` in database, you implement a seperate resolver for `postsCount` and return the `postsCount` according to the parent object. Example for complex object you already know, `posts` is one example. You're not restricted to resolve all scalar values in parent resolver and all complex objects in nested resolvers. Nested resolvers work for anything. 

Something as stupid as this is possible too:-
```
Query: {
  parentResolver: () => {
    return { name: "test" }
  }
},

Parent: {
  name: (parent) => {
    return parent.name.toUpperCase();
  }
}
```

The nested resolver here transforms the parent.name. This could've been done in the parent resolver itself, but this is just an example, you're free to do whatever you want with the values parent resolver returns in the nested resolvers. When graphql resolves parent.name field, it won't take the name you returned from parentResolver, it'll make use of the custom resolver you defined for parent.name. So, custom resolver take more precedence over the values you simply returned in parent resolver. If there's no custom resolver graphql will simply return the parent.name you returned in parent resolver. All it cares about is that the name and type of field you return matches with what's defined in the schema. 

Now whether you choose to implement data/errors object for each of these field(scalar or object) resolvers is up to you. This is something I've explained to you many times now. 

Do keep in mind, some scalar fields are database intensive to compute. I don't need to give an example here, you can probably think of many scalar fields like these. If you implement them in parent resolver, you're computing those scalar fields each time someone makes a request. That's bad for the database and server.

Implementing things in nested resolvers is one way of making sure computationally expensive fields don't get triggered unnecessarily. There's another way, it's called `look ahead` pattern. In this you take the graphql info object, parse it and only compute the specific fields asked by the clients. But this is high level complexity, because doing this manually would be braindead. Automating this is only possible if your graphql schema and database schema can be mapped and interconverted using some kind of compiler or something. This is something only tech giants would do.

Reply:
@xoldyckk like organizationMemberConnection and its `email` field for members be a good example to demonstrate the scalar field error in a nested resolver?

Reply:
@xoldyckk Ohh alright got it , thanks for explaining that so clearly and in detail.

Reply:
@kb-0311 The point is just to show that nested resolvers not something magical, they're needed for properly organizing the code and having a clean architecture, delegating computation of expensive or custom fields which are ephimeral i.e., only exist at the graphql level not the database. `email` is a good example for this, but you can use whatever you want like `bank_account_number`, `bank_balance`, `phone_number` etc.

Reply:
@kb-0311 When you resolve custom/complex fields within the parent resolver, you have the freedom to return errors related to those fields within the errors field of that parent resolver. It's possible to resolve your whole graphql schema and all the relations within a single resolver, but that's anti graphql. You see this is only different  from the root errors approach in that here errors will be type-safe, in root errors list they're not type-safe, you'll still be resolving errors for all fields/nested fields within the parent resolver, and that'll make a big list of possible errors that parent resolver can return.

Extracting data and errors into resolvers is a modular approach and isolated information of a field to that field's resolver, the resolvers can be scalar or complex objects. But this does result in the graphql schema not representing the relations directly, instead you have `{success,errors}` objects for many fields(scalar or complex) which looks weird.

Reply:
Oh yeah, I agree with you completely. In talawa, in the future, we can even introduce fields that are not necessarily defined cut to cut in the database but could be valuable data to be sent back to the client. The example you stated yourself included.

I could not agree more with the tradeoff points you made. Ultimately I will also need to talk to the client apps to discuss how the corresponding UI elements will be developed for this 6a approach too since a lot of the motive behind this was to allow more customization in errors within the apps.